thirtyinch/src/main/java/net/grandcentrix/thirtyinch/internal/PresenterSavior.java

```diff
@@ -32,36 +39,261 @@
+    public void free(final String presenterId, @NonNull final Activity activity) {
+        final ActivityScopedPresenters scope = getScope(activity);
+        if (scope != null) {
+            scope.remove(presenterId);
+            clearScopeWhenPossible(activity);
+        }
+
+        printRemainingStore();
     }
 
     @Override
     @Nullable
-    public TiPresenter recover(final String presenterId) {
-        return mPresenters.get(presenterId);
+    public TiPresenter recover(final String presenterId, @NonNull final Activity activity) {
+        final ActivityScopedPresenters scope = getScope(activity);
+        if (scope == null) {
+            return null;
+        }
+        return scope.get(presenterId);
     }
 
     @Override
-    public String save(@NonNull final TiPresenter presenter) {
+    public String save(@NonNull final TiPresenter presenter, @NonNull final Activity activity) {
         final String id = generateId(presenter);
         TiLog.v(TAG, "save presenter with id " + id + " " + presenter);
-        mPresenters.put(id, presenter);
+
+        final ActivityScopedPresenters scope = getScopeOrCreate(activity);
+        scope.save(id, presenter);
+        printRemainingStore();
+
         return id;
     }
 
+    /**
+     * the {@link Activity} is in terminal state and got finished. cleanup all presenters
+     */
+    void cleanupAfterFinish(@NonNull final Activity activity) {
+        TiLog.d(TAG, "cleanupAfterFinish() called with: activity = [" + activity + "]");
+
+        final String scopeId = mScopeIdForActivity.remove(activity);
+        final ActivityScopedPresenters scope = mScopes.remove(scopeId);
+
+        if (mScopeIdForActivity.isEmpty()) {
+            TiLog.v(TAG, "unregistering lifecycle callback");
+            if (mLifecycleCallbacks != null) {
+                activity.getApplication().unregisterActivityLifecycleCallbacks(mLifecycleCallbacks);
+            }
+        }
+
+        TiLog.d(TAG, "Activity is really finishing!");
+        if (scope != null) {
+            for (final Map.Entry<String, TiPresenter> entry : scope.getAllMappings()) {
+                final String key = entry.getKey();
+                final TiPresenter presenter = entry.getValue();
+                if (!presenter.isDestroyed()) {
+                    presenter.destroy();
+                }
+                scope.remove(key);
+            }
+        }
+
+        printRemainingStore();
+    }
+
+    @VisibleForTesting
+    /*package*/ int getPresenterCount() {
+
+        final ArrayList<TiPresenter> presenters = new ArrayList<>();
+        for (final Map.Entry<String, ActivityScopedPresenters> entry : mScopes.entrySet()) {
+            presenters.addAll(entry.getValue().getAll());
+        }
+        return presenters.size();
+    }
+
+    private synchronized void clearScopeWhenPossible(@NonNull final Activity activity) {
+        final String scopeId = mScopeIdForActivity.get(activity);
```

Comment:
As I already said (feels like 100 times :D) scope is not a good naming here.
Thinking about `activtiyId` 🤔